### PR TITLE
first draft of Azure devops agent module

### PR DESCRIPTION
**Pull request by nick-o**
this module lets you build azure devops agent images for AWS and Azure Cloud using packer and then deploy them into AWS and/or Azure Cloud using terraform.

First Draft

### CODE REVIEW

File: terraform-azure-devops-agent/packer/scripts/install-base.sh [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+0HDxo5pwZuWzuE6DVL42K8rQZfjbZYdSQeUuON5iavPjqIUU4ghFpn7Tj/RoD9KB
+USJnFSGTcNIPGRhxAibMM/GE7Hrf5HAxG3WRVYrdw08gy9LLJGcPCSTECock80nl
+A+Npu232V2UYa9XAnjWrCincwCZkbT2OG7IYPtfH00DMffdTUkN23NndRCqCssyf
+9hPSBV8Dm7PXdXd6lkJn9rEvr/b859Q6bd85H2K9pIgmKhkYD9ZkIAHDgTQ2J1T/
+OG3uVx6q+NX6VQWioDTAaBmXMvuSCZbcexr0i5SKUc+3mpRHhHF6voqDg/RE/FLs
+ztR/6eWns9sfGiAt2Ye9vvIsSn703yfcYX4l02woMTZ4d43M7mD3NA8cj/sGfnQN
+lh9xpuI9M+8ZHCYJ
+=VSVF
+-----END PGP PUBLIC KEY BLOCK-----
+-----BEGIN PGP PUBLIC KEY BLOCK-----
```
Comment (selamanse) [inline terraform-azure-devops-agent/packer/scripts/install-base.sh:162]: try and get zulu key from here: http://repos.azulsystems.com/RPM-GPG-KEY-azulsystems

File: terraform-azure-devops-agent/azure_vm.tf [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+resource "azurerm_subnet" "internal" {
+  name                 = "internal"
+  resource_group_name  = azurerm_resource_group.main.name
+  virtual_network_name = azurerm_virtual_network.main.name
+  address_prefixes     = ["10.0.2.0/24"]
+}
+
+resource "azurerm_subnet" "bastion" {
```
Comment (selamanse) [inline terraform-azure-devops-agent/azure_vm.tf:26]: do we need bastion for first draft?

File: terraform-azure-devops-agent/packer/bin/build-packer-aws-hcl.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  --output text)
+assert_non_zero "${VPC_ID}" "VPC_ID"
+# shellcheck disable=SC2155
+DD_API_KEY=$(fetch_secretsmanager_value arn:aws:secretsmanager:eu-central-1:553574040935:secret:datadog/apikey-TUu0ml)
+assert_non_zero "${DD_API_KEY}" "DD_API_KEY"
+SG_ID=$(aws ec2 create-security-group --group-name "packer-ssh-${TIMESTAMP}" --description "security group for packer" --vpc-id "${VPC_ID}" --output text)
+assert_non_zero "${SG_ID}" "SG_ID"
+trap "rm -f ${TMPFILE}; aws ec2 delete-security-group --group-id ${SG_ID}" EXIT
+aws ec2 authorize-security-group-ingress --group-id "${SG_ID}" --protocol tcp --port 22 --cidr 0.0.0.0/0 --output text
+#SG_IDS="$(aws ec2 describe-security-groups --filters Name=tag:Name,Values=webserver-reference,webserver_ref --query SecurityGroups[].GroupId --output text | tr '[:blank:]' ','),${SG_ID}"
```
Comment (selamanse) [inline terraform-azure-devops-agent/packer/bin/build-packer-aws-hcl.sh:43]: remove commented code

**Review by selamanse — CHANGES_REQUESTED** [review]

File: terraform-azure-devops-agent/packer/bin/build-packer-azure-hcl.sh [hunk 4]
```diff
@@ -0,0 +1,30 @@
+#!/bin/sh
+
+set -ae
+assert_non_zero() {
+  if [ -z "${1}" ]; then
+    echo "${2} is empty!"
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/bin/build-packer-azure-hcl.sh:6]: this is not needed in the azure image build script at all so I have removed it here. It is still required in the AWS AMI build script so will remain in there. The syntax is right here, the usage of this pre-existing function is: 
```
assert_non_zero "${VAR}" "VAR"
```
basically meaning the value and name of the var need to be passed in separately to get useful output.

File: terraform-azure-devops-agent/packer/devops-agent-aws.pkr.hcl [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    filters = {
+      "vpc-id" : "${var.vpc_id}",
+      "tag:tier" : "public"
+    }
+    random = true
+  }
+  tags = {
+    "Name" : "azure-devops",
+    "service" : "azure-devops",
+    "team" : "thundercats"
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/devops-agent-aws.pkr.hcl]: Done

File: terraform-azure-devops-agent/packer/devops-agent-aws.pkr.hcl [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  var_retrieval = templatefile("${path.root}/templates/azure-vars.pkrtpl.hcl", {
+    SECRET_ID = var.devops_org_token_secret_arn
+  })
+}
+
+source "file" "example" {
+  content = "example content"
+}
+
+source "amazon-ebs" "ssm-example" {
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/devops-agent-aws.pkr.hcl]: done

File: terraform-azure-devops-agent/packer/devops-agent-aws.pkr.hcl [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  instance_type               = "m5.large"
+  region                      = var.region
+  source_ami                  = data.amazon-ami.ubuntu.id
+  ssh_username                = "ubuntu"
+  ssh_clear_authorized_keys   = true
+  encrypt_boot                = false
+  associate_public_ip_address = true
+  communicator                = "ssh"
+  security_group_ids          = [var.sg_id]
+  # temporary_iam_instance_profile_policy_document {
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/devops-agent-aws.pkr.hcl]: removed comment

File: terraform-azure-devops-agent/packer/bin/build-packer-aws-hcl.sh (hunk 3 above)
Comment (nick-o) [reply terraform-azure-devops-agent/packer/bin/build-packer-aws-hcl.sh]: done

File: terraform-azure-devops-agent/packer/scripts/install-base.sh [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+sudo DEBIAN_FRONTEND=noninteractive sudo apt-get -yq install gnupg curl 
+sudo apt-key adv \
+  --keyserver hkp://keyserver.ubuntu.com:80 \
+  --recv-keys 0xB1998361219BD9C9
+curl -O https://cdn.azul.com/zulu/bin/zulu-repo_1.0.0-3_all.deb
+
+# add the datadog key: A2923DFF56EDA6E76E55E492D3A80E30382E94DE
+# add the zulu key: 0xB1998361219BD9C9
+sudo apt-key adv --import <<EOF
+-----BEGIN PGP PUBLIC KEY BLOCK-----
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/scripts/install-base.sh:19]: that's done

File: terraform-azure-devops-agent/azure_image.tf [hunk 9]
```diff
@@ -0,0 +1,5 @@
+data "azurerm_image" "devops_agent_packer" {
+  name_regex          = "devops-agent.*"
+  resource_group_name = "crm-shared-rg"
```
Comment (selamanse) [inline terraform-azure-devops-agent/azure_image.tf:3]: use variable for rg

File: terraform-azure-devops-agent/packer/scripts/install-agent.sh [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+#
+#if [ -z "${LINUX_AGENT_URL}" ]; then
+#  echo "missing mandatory env var: LINUX_AGENT_URL"
+#  exit 1
+#fi
+
+sudo mkdir -p /src
+sudo chown ubuntu /src
+
+LINUX_AGENT_URL="https://vstsagentpackage.azureedge.net/agent/2.194.0/vsts-agent-linux-x64-2.194.0.tar.gz",
```
[the remark this answers is not in the document]
Comment (selamanse) [reply terraform-azure-devops-agent/packer/scripts/install-agent.sh:13]: agent download should be separated to phase packer build

[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/scripts/install-agent.sh]: agent version is now configurabe via `DEVOPS_AGENT_VERSION` env var, will use default of 2.194.0 if aforementioned env var is not set.

File: terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+mkdir -p /src
+chown ubuntu /src
+chgrp ubuntu /src
+
+${RETRIEVE_PARAMETERS}
+
+LINUX_AGENT_URL="https://vstsagentpackage.azureedge.net/agent/2.194.0/vsts-agent-linux-x64-2.194.0.tar.gz"
+DEVOPS_ORG_URL="https://dev.azure.com/EDDI-CRM"
+DEVOPS_AGENT_POOL="aws-by-thundercats"
+NAME=best-devops-agent-ever-$INSTANCE_NAME
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl]: done

File: terraform-azure-devops-agent/azure_vm.tf (hunk 2 above)
Comment (nick-o) [reply terraform-azure-devops-agent/azure_vm.tf:67]: really just the subnet for it and a NSG, not the actual bastion. But can get rid, just saved me some time

File: terraform-azure-devops-agent/packer/scripts/install-agent.sh [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+DEVOPS_ORG_URL="https://dev.azure.com/EDDI-CRM",
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/scripts/install-agent.sh]: done

File: terraform-azure-devops-agent/packer/scripts/install-agent.sh [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+DEVOPS_AGENT_POOL="aws-by-thundercats",
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/scripts/install-agent.sh]: done

File: terraform-azure-devops-agent/packer/scripts/install-agent.sh [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+DEVOPS_ORG_TOKEN="TOKEEEEN"
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/scripts/install-agent.sh]: token is pulled from Secretsmanager/KeyVault

File: terraform-azure-devops-agent/packer/templates/azure-vars.pkrtpl.hcl [hunk 15]
```diff
@@ -0,0 +1,7 @@
+pip3 install azure-keyvault==1.1.0
+
+INSTANCE_NAME=$(hostname)
+az login --identity --allow-no-subscriptions
+az keyvault secret download --vault-name ${VAULT_NAME} --name ${SECRET_NAME} --file secret.txt
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/templates/azure-vars.pkrtpl.hcl]: done

File: terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+#  exit 1
+#fi
+
+mkdir -p /src
+chown ubuntu /src
+chgrp ubuntu /src
+
+${RETRIEVE_PARAMETERS}
+
+LINUX_AGENT_URL="https://vstsagentpackage.azureedge.net/agent/2.194.0/vsts-agent-linux-x64-2.194.0.tar.gz"
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl]: done

File: terraform-azure-devops-agent/versions.tf [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+  required_version = ">= 1.0.0"
+
+  
+  required_providers {
+   
+  }
+}
+
+provider "aws" {
+  region = "eu-central-1"
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/versions.tf]: provider block has been removed as this is a module

File: terraform-azure-devops-agent/packer/scripts/install-base.sh (hunk 1 above)
Comment (nick-o) [reply terraform-azure-devops-agent/packer/scripts/install-base.sh]: done

File: terraform-azure-devops-agent/azure_vm.tf (hunk 2 above)
Comment (nick-o) [reply terraform-azure-devops-agent/azure_vm.tf]: removed

File: terraform-azure-devops-agent/azure_image.tf (hunk 9 above)
Comment (nick-o) [reply terraform-azure-devops-agent/azure_image.tf]: done

File: terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+DEVOPS_ORG_URL="https://dev.azure.com/EDDI-CRM"
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl]: done

File: terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+DEVOPS_AGENT_POOL="aws-by-thundercats"
```
[the remark this answers is not in the document]
Comment (nick-o) [reply terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl]: done

File: .github/workflows/terraform-azure-devops-agent-packer.yml [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+      uses: selamanse/packer-github-actions@master
+      with:
+        command: validate
+        arguments: -syntax-only
+        target: ${{ env.WORKDIR }}/${{ matrix.pkr }}
+
+  packer-build:
+    name: 'build-packer-image'
+    runs-on: ubuntu-latest
+    #needs: [validate]
```
Comment (selamanse) [inline .github/workflows/terraform-azure-devops-agent-packer.yml:91]: Uncomment to build only after verify

File: .github/workflows/terraform-azure-devops-agent-packer.yml [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+    #if: success() && github.ref == 'refs/heads/master'
```
Comment (selamanse) [inline .github/workflows/terraform-azure-devops-agent-packer.yml:92]: Uncomment to build on master

File: terraform-azure-devops-agent/packer/scripts/install-base.sh [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+sudo DEBIAN_FRONTEND=noninteractive sudo apt-get -yq install gnupg curl 
+sudo apt-key adv \
+  --keyserver hkp://keyserver.ubuntu.com:80 \
+  --recv-keys 0xB1998361219BD9C9
+curl -O https://cdn.azul.com/zulu/bin/zulu-repo_1.0.0-3_all.deb
+
+# add the zulu key: 0xB1998361219BD9C9
+sudo su - -c "curl http://repos.azulsystems.com/RPM-GPG-KEY-azulsystems | apt-key adv --import -"
+# sudo apt-key adv --import <<EOF
+# -----BEGIN PGP PUBLIC KEY BLOCK-----
```
Comment (selamanse) [inline terraform-azure-devops-agent/packer/scripts/install-base.sh:20]: Why not remove this comment? Is it still necessary ?

File: terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl [hunk 23]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+set -ex
+
+source /etc/environment
+env
+
+${RETRIEVE_PARAMETERS}
+
+#DEVOPS_ORG_URL= #"https://dev.azure.com/EDDI-CRM"
```
Comment (selamanse) [inline terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl:9]: Remove pls

File: terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+#DEVOPS_AGENT_POOL="aws-by-thundercats"
```
Comment (selamanse) [inline terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl:10]: Remove pls

**Review by selamanse — CHANGES_REQUESTED** [review]
Only a couple of minors left ;)

File: .github/workflows/terraform-azure-devops-agent-packer.yml (hunk 20 above)
Comment (nick-o) [reply .github/workflows/terraform-azure-devops-agent-packer.yml]: done

File: terraform-azure-devops-agent/packer/scripts/install-base.sh (hunk 22 above)
Comment (nick-o) [reply terraform-azure-devops-agent/packer/scripts/install-base.sh]: done

File: .github/workflows/terraform-azure-devops-agent-packer.yml (hunk 21 above)
Comment (nick-o) [reply .github/workflows/terraform-azure-devops-agent-packer.yml]: done

File: terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl (hunk 23 above)
Comment (nick-o) [reply terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl]: done

File: terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl (hunk 24 above)
Comment (nick-o) [reply terraform-azure-devops-agent/packer/templates/install-agent.pkrtpl.hcl]: done

**Review by selamanse — APPROVED** [review]

**Pull request merged**